### PR TITLE
Feature/tri 986 implement edc catalog caching

### PR DESCRIPTION
**Pull request by ds-jhartmann**

### CODE REVIEW

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            catalogCache.get(key).removeIf(catalogItem -> catalogItem.getValidUntil().isBefore(Instant.now()));
+        }
+    }
+
+    private void removeOldestCacheValues(final int listSize) {
+        final List<CatalogItem> oldestCatalogItems = catalogCache.values()
+                                                                 .stream()
+                                                                 .flatMap(List::stream)
+                                                                 .sorted((o1, o2) -> (int) (
+                                                                         o1.getValidUntil().toEpochMilli() - o2.getValidUntil().toEpochMilli()))
```
Comment (ds-jkreutzfeld) [inline irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java:135]: I think you can just use Instant.compareTo

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        if (cacheConfig.isEnabled()) {
+            List<CatalogItem> catalogItems = new ArrayList<>();
+            if (catalogCache.containsKey(connectorUrl)) {
+                catalogItems = catalogCache.get(connectorUrl);
+            }
+            final int listSize = catalog.size();
+            if (!cacheHasSpaceLeft(listSize)) {
+                removeOldestCacheValues(listSize);
+            }
+            catalogItems.addAll(catalog.stream().map(this::updateTTL).toList());
```
Comment (ds-jkreutzfeld) [inline irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java:105]: What happens if the number of items in the catalog is larger than the cache size limit?

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/EDCCatalogFacade.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ */
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class EDCCatalogFacade {
+
+    private final EdcControlPlaneClient controlPlaneClient;
+    private final EdcConfiguration config;
+
+    public List<CatalogItem> getCatalog(final String connectorUrl, final String target) {
```
Comment (ds-jkreutzfeld) [inline irs-api/src/main/java/org/eclipse/tractusx/irs/edc/EDCCatalogFacade.java:52]: It's not clear to me from the method name that we are not retrieving the whole catalog, but instead all items of the catalog until we find a match.

**Review by ds-jkreutzfeld — CHANGES_REQUESTED** [review]

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java (hunk 2 above)
Comment (ds-jhartmann) [reply irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java]: I changed the implementation to limit the max amount of at once cached items to the max configured cache size.

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/EDCCatalogFacade.java (hunk 3 above)
Comment (ds-jhartmann) [reply irs-api/src/main/java/org/eclipse/tractusx/irs/edc/EDCCatalogFacade.java]: Changed method name and added JavaDoc.

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java (hunk 1 above)
Comment (ds-jhartmann) [reply irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java]: Done

File: irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        Optional<CatalogItem> catalogItem = getItemFromCache(connectorUrl, target);
+        if (catalogItem.isPresent()) {
+            final CatalogItem item = catalogItem.get();
+            log.info("Retrieved Item from cache: '{}'", item);
+            return item;
+        } else {
+            log.info("Retrieving Catalog from connector '{}'", connectorUrl);
+            catalogItem = getOfferFromCatalog(connectorUrl, target);
+            catalogItem.ifPresent(item -> log.info("Retrieved Item from connector: '{}'", item));
+            return catalogItem.orElseThrow();
```
[the remark this answers is not in the document]
Comment (ds-jhartmann) [reply irs-api/src/main/java/org/eclipse/tractusx/irs/edc/CatalogCache.java]: Done

**Review by ds-jkreutzfeld — DISMISSED** [review]

**Review by ds-jkreutzfeld — APPROVED** [review]

**Pull request merged**